### PR TITLE
Add dataset readers

### PR DESCRIPTION
**Pull request by JohnGiorgi**
# Overview

This pull request adds `torchtext` dataset readers for three major dataset types (document classification, sequence labelling, and NLI). Adding new dataset loaders should be straightforward:

1. Subclass `DatasetReader`
2. Define `textual_to_iterator` method which defines the torchtext fields, calls the parents `textual_to_iterator`, builds the vocabulary (`FIELD.build_vocab`) and return the iterators.

## TODO
- [x] Allow `DocumentClassificationDatasetReader` to handle both single label and multi label classification (@nicolasahar).
~~- [ ] Fix/test `SequenceLabellingClassificationHead` (@JohnGiorgi)~~
- [x] Add `RelationClassificationDatasetReader` (@JohnGiorgi)
~~- [ ] Prevent blinded entities from being split in RC tasks (@JohnGiorgi)~~
~~- [ ] Unit test all dataset readers (@JohnGiorgi)~~
- [x] Our use of `fix_length` is causing all batches to be of the same size (e.g. the bucket iterator is pointless). Try passing `max_len` to `encode` instead. (@JohnGiorgi, @nicolasahar)

### CODE REVIEW

**Comment by JohnGiorgi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Coverage increased per file
===========================
+ drbert/utils/data_utils.py  50
+ drbert/heads.py  90
+ drbert/utils/eval_utils.py  20
+ drbert/data/dataset_readers.py  70
         

Complexity increasing per file
==============================
- drbert/utils/data_utils.py  7
- drbert/heads.py  3
- drbert/utils/eval_utils.py  12
- drbert/run.py  37
- drbert/data/dataset_readers.py  4
         

Clones added
============
- drbert/utils/data_utils.py  2
- drbert/data/dataset_readers.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662&bid=14661663)

File: drbert/data/dataset_readers.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            sequential examples.
+        format (str): The format of the data file. One of “CSV”, “TSV”, or “JSON” (case-insensitive).
+        batch_sizes (tuple): Optional, tuple of batch sizes to use for the different splits, or None
+            to use the same batch_size for all splits. Defaults to None.
+        maxlen (int): Optional, a fixed length that all examples using this field will be padded to,
+            for flexible sequence lengths. Default: 512.
+        lower_case: (bool): Optional, True if text should be lower cased, False if not. Defaults to
+            False.
+    """
+    def __init__(self, path, partitions, tokenizer, format='json', batch_sizes=None, maxlen=512,
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        maxlen (int): Optional, a fixed length that all examples using this field will be padded to,
+            for flexible sequence lengths. Default: 512.
+        lower_case: (bool): Optional, True if text should be lower cased, False if not. Defaults to
+            False.
+
+    Raises:
+        ValueError: If batch_sizes is not a tuple or None.
+        ValueError: If batch_sizes is not None and `len(batch_sizes) != len(partitions)`.
+    """
+    def __init__(self, path, partitions, tokenizer, format, batch_sizes=None, maxlen=512,
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            sequential examples.
+        format (str): The format of the data file. One of “CSV”, “TSV”, or “JSON” (case-insensitive).
+        batch_sizes (tuple): Optional, tuple of batch sizes to use for the different splits, or None
+            to use the same batch_size for all splits. Defaults to None.
+        maxlen (int): Optional, a fixed length that all examples using this field will be padded to,
+            for flexible sequence lengths. Default: 512.
+        lower_case: (bool): Optional, True if text should be lower cased, False if not. Defaults to
+            False.
+    """
+    def __init__(self, path, partitions, tokenizer, format='csv', batch_sizes=None, maxlen=512,
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            tokenize=self.tokenizer.encode,
+            batch_first=True,
+            pad_token=self.tokenizer.pad_token_id,
+            unk_token=self.tokenizer.unk_token_id
+        )
+        LABEL = data.LabelField()
+
+        fields = [('text', TEXT), ('label', LABEL)]
+
+        splits, iterators = super(NLIDatasetReader, self).textual_to_iterator(fields)
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Bad first argument 'NLIDatasetReader' given to super()](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    """
+    def __init__(self, path, partitions, tokenizer, format='json', batch_sizes=None, maxlen=512,
+                 lower_case=False):
+        super(NLIDatasetReader, self).__init__(
+            path, partitions, tokenizer, format, batch_sizes, maxlen, lower_case,
+            # Sort examples according to length of premise and hypothesis
+            sort_key=lambda x: data.interleave_keys(len(x.premise), len(x.hypothesis))
+        )
+
+    def textual_to_iterator(self):
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Parameters differ from overridden 'textual_to_iterator' method](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    """
+    def __init__(self, path, partitions, tokenizer, batch_sizes=None, fix_length=512, lower=False):
+        super(NLIDatasetReader, self).__init__(
+            path=path, partitions=partitions, tokenizer=tokenizer, format='JSON', skip_header=False,
+            batch_sizes=batch_sizes, fix_length=fix_length, lower=lower,
+            # Sort examples according to length of premise and hypothesis
+            sort_key=lambda x: data.interleave_keys(len(x.premise), len(x.hypothesis))
+        )
+
+    def textual_to_iterator(self):
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Parameters differ from overridden 'textual_to_iterator' method](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

**Comment by JohnGiorgi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- drbert/data/dataset_readers.py  4
- drbert/run.py  37
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662&bid=14661663)

File: drbert/data/dataset_readers.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+            False.
+        sort_key (function): A key to use for sorting examples in order to batch together examples
+            with similar lengths and minimize padding. The sort_key provided to the Iterator
+            constructor overrides the sort_key attribute of the Dataset, or defers to it if None.
+
+    Raises:
+        ValueError: If batch_sizes is not a tuple or None.
+        ValueError: If batch_sizes is not None and `len(batch_sizes) != len(partitions)`.
+    """
+    def __init__(self, path, partitions, tokenizer, format, skip_header=False, batch_sizes=None,
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

**Comment by JohnGiorgi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- drbert/data/dataset_readers.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662&bid=14661663)

File: drbert/data/dataset_readers.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+    """
+    def __init__(self, path, partitions, tokenizer, batch_sizes=None, lower=False):
+        super(NLIDatasetReader, self).__init__(
+            path=path, partitions=partitions, tokenizer=tokenizer, format='JSON', skip_header=False,
+            batch_sizes=batch_sizes, lower=lower,
+            # Sort examples according to length of premise and hypothesis
+            sort_key=lambda x: data.interleave_keys(len(x.premise), len(x.hypothesis))
+        )
+
+    def textual_to_iterator(self):
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Parameters differ from overridden 'textual_to_iterator' method](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+        tokenizer (transformers.PreTrainedTokenizer): The function used to tokenize strings into
+            sequential examples.
+        batch_sizes (tuple): Optional, tuple of batch sizes to use for the different splits, or None
+            to use the same batch_size for all splits. Defaults to None.
+        lower: (bool): Optional, True if text should be lower cased, False if not. Defaults to
+            False.
+        multi_label_map (dict or list): dict or list containing all fields in the label space as per the jsonl file (for multi-label document classification only). If None, will assume single label document classification. Defaults to None 
+    """
+    def __init__(self, path, partitions, tokenizer, batch_sizes=None, lower=False, multi_label_map=None):
+        super(DocumentClassificationDatasetReader, self).__init__(
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Bad first argument 'DocumentClassificationDatasetReader' given to super()](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/run.py [hunk 10]
```diff
@@ -57,8 +68,6 @@ def prepare_optimizer_and_scheduler(args, model, t_total):
     decay_blacklist = {'LayerNorm.bias', 'LayerNorm.weight'}
 
     decay, no_decay = [], []
-
-    for name, param in model.named_parameters():
```
Comment (nicolasahar) [inline drbert/run.py]: @JohnGiorgi I think this line got deleted by accident

Comment (JohnGiorgi) [reply drbert/run.py]: Yes, nice catch!

**Comment by JohnGiorgi** [comment]
@nicolasahar Updating your dataset reader a little. There is no dict or list provided by the user anymore. Instead, any field that isn't in `blacklist` (which is currently just `{'test', 'id'}`) is considered a label field.

**Comment by JohnGiorgi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- drbert/data/dataset_readers.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662&bid=14661663)

File: drbert/data/dataset_readers.py [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+    """
+    def __init__(self, path, partitions, tokenizer, batch_sizes=None, lower=False):
+        super(NLIDatasetReader, self).__init__(
+            path=path, partitions=partitions, tokenizer=tokenizer, format='JSON', skip_header=False,
+            batch_sizes=batch_sizes, lower=lower,
+            # Sort examples according to length of premise and hypothesis
+            sort_key=lambda x: data.interleave_keys(len(x.premise), len(x.hypothesis))
+        )
+
+    def textual_to_iterator(self):
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Parameters differ from overridden 'textual_to_iterator' method](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+            False.
+        sort_key (function): A key to use for sorting examples in order to batch together examples
+            with similar lengths and minimize padding. The sort_key provided to the Iterator
+            constructor overrides the sort_key attribute of the Dataset, or defers to it if None.
+
+    Raises:
+        ValueError: If batch_sizes is not a tuple or None.
+        ValueError: If batch_sizes is not None and `len(batch_sizes) != len(partitions)`.
+    """
+    def __init__(self, path, partitions, tokenizer, format, skip_header=False, batch_sizes=None,
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

File: drbert/data/dataset_readers.py [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

**Comment by JohnGiorgi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- drbert/data/dataset_readers.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662&bid=14661663)

File: drbert/data/dataset_readers.py [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+    """
+    def __init__(self, path, partitions, tokenizer, batch_sizes=None, lower=False):
+        super(NLIDatasetReader, self).__init__(
+            path=path, partitions=partitions, tokenizer=tokenizer, format='JSON', skip_header=False,
+            batch_sizes=batch_sizes, lower=lower,
+            # Sort examples according to length of premise and hypothesis
+            sort_key=datasets.nli.NLIDataset.sort_key
+        )
+
+    def textual_to_iterator(self):
```
Comment (JohnGiorgi) [inline drbert/data/dataset_readers.py]: Codacy Issue found: [Parameters differ from overridden 'textual_to_iterator' method](https://app.codacy.com/app/JohnGiorgi/drbert/pullRequest?prid=4302662)

**Pull request merged**